### PR TITLE
Template Locking: readonly / disable editing attributes

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This PR adds an option to the `<InnerBlocks>` `templateLock` prop to disable *all* editing inside a template (block attributes).

Unfortunately there is already an option `all`, which doesn't really locks *all* editing. It only disables block inserting, moving and removing.

I'm not 100% sure if adding another option is the best thing. Perhaps we should have a prop that allows an array of allowed (or disallowed) operations. Something like `lock = [ 'insert', 'delete', 'move', 'attributes' ]` where `lock = true` disables all operations. Thoughts?

~~The way the attributes are locked is by using the `save` function instead of the `edit` function.~~ Decided to use `edit`, so it's guaranteed to exist. A prop `isReadOnly` will be passed down, so a block could disable attribute editing controls. Block controls that we provide, such as `RichText`, `PlainText`, `Placeholder` etc. will be disabled by default. The block toolbar and inspector are aso disabled. To ensure so attributes get update by a block, `setAttributes` will not do anything.

## How has this been tested?

Enable the CPT Locking test plugin. You should not be able to edit anything.

For further testing (with content) you could alter e.g. the group block's `edit` function and adjust `InnerBlocks`.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by chrisvanpatten** [comment]
Yes yes yes!

> I'm not 100% sure if adding another option is the best thing. Perhaps we should have a prop that allows an array of allowed (or disallowed) operations. Something like lock = [ 'insert', 'delete', 'move', 'attributes' ] where lock = true disables all operations. Thoughts?

I agree completely! I actually have a few random proposals for this over in #8112. I think a more versatile option for `templateLock`ing would be wildly useful.

**Comment by ellatrix** [comment]
@chrisvanpatten Nice, I think that's what we should do then.

**Comment by chrisvanpatten** [comment]
@ellatrix Is there anything I can do to help unblock/move this forward?

**Comment by ellatrix** [comment]
Review? 🙂

**Comment by ellatrix** [comment]
I'll rebase the PR

**Comment by epiqueras** [comment]
Yes, I think this array form of `templateLock` is much better and I like how we are still able to keep backwards compatibility with the string form and have our components infer the lock from their context.

File: docs/designers-developers/developers/block-api/block-templates.md [hunk 1]
```diff
@@ -105,10 +105,17 @@ function myplugin_register_template() {
 add_action( 'init', 'myplugin_register_template' );
 ```
 
-*Options:*
+*Array options:*
 
-- `all` — prevents all operations. It is not possible to insert new blocks, move existing blocks, or delete blocks.
-- `insert` — prevents inserting or removing blocks, but allows moving existing blocks.
+- `'insert'` — prevents new blocks from being inserted.
+- `'move'` — prevents blocks form being moved.
```
Comment (mcsf) [inline docs/designers-developers/developers/block-api/block-templates.md]:
```suggestion
- `'move'` — prevents blocks from being moved.
```

**Comment by mcsf** [comment]
> I'll rebase the PR, but would be good to know if this is a good direction

I think so, @ellatrix.

**Comment by richrd** [comment]
This would be great to have! I'm prototyping a slider block which needs to disable editing of the inner blocks when the slider is in preview mode. I tried `templateLock: "all"` but that still allows tabbing into the blocks within the slides. That causes all the edit controls to appear which breaks the slider layout.

I haven't found a solution yet but it looks like this PR would solve my problem :+1:

**Comment by guarani** [comment]
Looks like this needs to be rebased with `master` and conflicts resolved to be testable on mobile since this PR pre-dates the monorepo changes.

**Pull request closed without merging**